src/convolution/convolution_2d.jl

```diff
@@ -0,0 +1,1191 @@
+
+# ╔═╡ 9be00bec-59a5-478b-ada2-854f7a52d66e
+md"""To keep things simple, we will only deal with black and white images for now. So, for a black and white image, each pixel is a number between 0 and 1. 
+
+Try it out! Move the slider around to set the pixel below to get a different shade of gray"""
+
+# ╔═╡ c4da04ab-f7e0-46fd-b352-e518e4733608
+g_slider = @bind g PlutoUI.Slider(0:0.05:1)
+
+# ╔═╡ 96a4b35a-5a3a-4ad0-9ffe-306db46d1c03
+pixel = Gray.(g)
+
+# ╔═╡ 3f830a9b-3a78-4d18-ae3e-933778402a25
+md"""Now we can generate a 5x5 table of random pixels and combine them together and we have already our first image!
+"""
+
+# ╔═╡ 98919284-8099-43c8-b123-d1476d27a88d
+img = rand(Gray, 5, 5)
+
+# ╔═╡ eb983abe-5899-431b-84a1-7d06715d75c4
+md"""
+Now that you know how images are built. Let's get to the fun part. Below is a dropdown with popular test images. Choose the one you like and let's start learning about filters.
+"""
+
+# ╔═╡ 4edbfca5-cb25-4093-9f4d-5cfa2aa8d77f
+@bind test_img_selected Select([
+	testimage("coffee") => "Coffee", 
+	testimage("house") => "House",
+	testimage("pirate") => "Pirate",
+	testimage("peppers") => "Peppers",
+	testimage("mandrill") => "Mandrill"
+	])
+
+# ╔═╡ ef0bea69-6dfd-4526-a96e-6607771660fe
+test_img = Gray.(test_img_selected)
+
+# ╔═╡ ebb63415-2738-47f3-b0ef-0fd6e6bac259
+md"""
+## Using Filter 
+
+Now to make things more interesting, we will add some code that applies a filter to this image.
+
+This code basically goes over all the pixels in the image and applies the mathematical operation of convolution, that is the values of the pixel and its surrounding pixels are multipled by the value of the filter matrix. 
+
+However, it's alright if you don't understand exactly how the code works."""
+
+# ╔═╡ 44d594ff-a25b-4455-be56-863345c67b68
+function convolve(img, filter)
+
+	# reads the sizes of the image and the filter
+    img_row, img_col = size(img)
+    filter_row, filter_col = size(filter)
+
+	# initialize the final image with zeros and the same size as the original image
+    filtered = zeros(img_row-filter_row+1, img_col-filter_col+1)
+    filtered_row, filtered_col = size(filtered)
+
+	# apply the filter to each pixel in the image by iterating over it, like iterating over a table/2D array/matrix
+    for i in 1:filtered_row
+        for j in 1:filtered_col
+            filtered[i,j] = sum(img[i:i+filter_row-1,j:j+filter_col-1].*filter)
+        end
+    end
+  
+    return filtered
+end
+
+# ╔═╡ d752805d-ba2c-4133-a556-46a1da734557
+md"""Now let's start with a simple filter. This is called a "Box Blur" and it has the simple task of blurring the image.
+
+Can you see it ?"""
+
+# ╔═╡ d729f8b4-05e2-4863-b8c4-39b37646c36b
+test_filter = [0.111 0.111 0.111; 0.111 0.111 0.111; 0.111 0.111 0.111]
+
+# ╔═╡ 8ae0248f-6eba-4941-85f4-b21be3c7e725
+let
+	height, width = size(test_img)
+	filtered_img = Gray.(convolve(test_img, test_filter))
+	padded_img = test_img[2:height-1, 2:width-1]
+	[padded_img filtered_img]
+end
+
```

Comment:
Again, I would add some extra explanation here. What is the matrix doing? Here is my suggestion:

```md
What's happening? For each pixel, the new value is calculated as the sum of:
- The value of the pixel multiplied by the center value in the matrix
- The value of the pixel to the top left of the pixel, multiplied by the top left value in the filter matrix
- The value of the pixel to to the top, multiplied by the center top value in the  filter matrix
- The value of the pixel to the left, multiplied by the center left value in the filter matrix
- Etc...

In the box blur filter, all the filter values are the same. So for each pixel, the new value is just the average value of the pixel and its neighbours!
```